### ISSUE TITLE
Automatically imply type for JSONExtract -- JSONExtractImplyType(data, key)

### ISSUE THREAD

**Issue by adamsommer** [ClickHouse/ClickHouse#44646]
I'd like to be able to run `JSONExtractImplyType` on a JSON column in Clickhouse to effectively leverage the combination of `JSONExtract` and `JSONType`

**Use case**

We have lots of data stored in a single string column containing JSON called `data`. The users of our platform define send us arbitrary JSON, so the data we store for each user is very different.

We allow our users to run arbitrary SQL queries over our data. 

When a user with `userId=123` writes 
```
SELECT * FROM events WHERE "someRandomKey" = 5;
```
we convert that user input to 
```
SELECT * FROM events WHERE JSONExtractRaw(data, "someRandomKey") = 5 AND "userId" = 123;
```
on the backend. This allows our users to make really powerful queries of data without knowing exactly how we are storing data in the back. 

Because we don't know ahead of time that `someRandomKey` is a numerical value, JSONExtractRaw pulls out a string, not a number, and the query fails. Similarly, if the user runs 
```
SELECT sum(someRandomKey) FROM events
```
the query fails. 

When the user sends {"someRandomKey": 5} (where 5 is a number) as opposed to {"someOtherKey": "5"} (where "5" is a string), I would assume that Clickhouse should be able to imply that the value is a numerical value or a string value, since that's what's in the JSON. Most other JSON libraries can make similar inferences. 

**Describe the solution you'd like**

I would like a method like `JSONExtractImplyType(data, key)` which implies whether `key` corresponds to a **string**, **numerical**, or **boolean** value (just like `JSONType` would infer). This would allow for queries like the following to work as you'd expect: 
```
SELECT sum(JSONExtractImplyType(data, numericalKey)) FROM events
SELECT count(*) FROM events WHERE JSONExtractImplyType(data, booleanKey))
...
```

**Describe alternatives you've considered**

I've tried combining `JSONExtract(data, key, JSONType(data, key))` but I get errors, so this doesn't work. 

Thanks!!

**Comment by den-crane** [comment]
> AND "userId" = 123;
>When a user with userId=123 writes

How is it relevant to the question?

**Comment by den-crane** [comment]
The problem here that SQL theory assumes that column datatypes are known BEFORE a query execution.

You have to use functions OrNull / OrZero, like `toFloat64OrNull` 

```sql
SELECT
    '{"someRandomKey":5}' AS j,
    JSONExtractRaw(j, 'someRandomKey') AS val,
    JSONType(j, 'someRandomKey') AS type,
    toFloat64OrNull(val) AS value_float64

Query id: 1ed0367b-3623-4156-9ea6-ce4222541fe0

┌─j───────────────────┬─val─┬─type──┬─value_float64─┐
│ {"someRandomKey":5} │ 5   │ Int64 │             5 │
└─────────────────────┴─────┴───────┴───────────────┘
```

**Comment by den-crane** [comment]
It's not a problem to create such function `JSONExtractImplyType`. 
The problem that such function is impossible to use with relational (SQL) dbms in SQL queries.

**Comment by adamsommer** [comment]
> AND "userId" = 123;
> When a user with userId=123 writes

This is simply to illustrate the multi-tenant nature of our data store — different users store data in our platform, and we translate their SQL-like queries into Clickhouse queries specific to their data

**Comment by adamsommer** [comment]
> It's not a problem to create such function `JSONExtractImplyType`.
> The problem that such function is impossible to use with relational (SQL) dbms in SQL queries.

Yes, SQL does know column types ahead of time. But _users_ of SQL do not need to think about data types when executing queries. 

`SELECT SUM(column) FROM table`
A user does not need to know whether `column` is an int, a float, a bigint… they can just make queries. If the column is a string, they'll get an error, but they don’t need to specify anything about the column type at query-time

I suppose ultimately what I’m asking for is a JSON query mechanism where the column type doesn’t need to be specified _at query time_

**Comment by UnamedRus** [comment]
Fixed in pr https://github.com/ClickHouse/ClickHouse/pull/41502
